### DISCUSSION TITLE
How to set console log level?

### DISCUSSION THREAD

**Discussion by retog** [denoland/deno#14143] [Q&A]
I assumed that setting the log-level to info would prevent messages passed to `console.debug` from appearing, however this doesn't seem to be the case:

```
$ echo 'console.debug("hi") ' | deno -L info run -
hi
```
Is there away to have only messages passed to `console.info` but not those passed to `console.debug`?

**Comment by kitsonk** [comment]
`deno -L` flag is for internal logging of Deno.  `console.*` is for _userland_ code.  We should make it clearer in the _help_ what logging it applies to.

`console.*` behaves a lot like it does in a browser. In a browser, `console.log()`, `console.info()`, `console.warn()`, `console.error()` change the "display" of the errors and advanced GUI allow filtering, but browsers do not allow "out of the box" suppression, nor does Node.js.  Node.js and Deno both send `console.error()` and `console.warn()` to stderr instead of stdout.

Adding out of the box suppression would be "surprising" to the established _de facto_ precedent established.

Because the _primitive_ logging is often not sufficient for a lot of workloads, people tend to use a logging framework.  There is a [logging framework](https://deno.land/std/log) as part of `std` library, but I think the larger community is split on its usefulness.  There are other [logging frameworks](https://deno.land/x/?query=log) available on `deno.land/x` that might be suitable as well.

If you are dealing with 3rd part code, then you might want to consider either patching the global `console` yourself, or use a package that helps you do that.

In the end, Deno treats `console.*` as a userland construct/API and doesn't _interfere_ with it.

**Reply by retog** [reply]
Thanks @kitsonk.

As you mention some GUIs allow filtering, isn't that the whole point of differentiating the log levels? How can a user of the command-line deno tools benefit from the debug/info distinction the developer made? Even in VSCode I see no way to filter the log-level.

I've been using the logging framework in std, but his has the big disadvantage that I don't have links to the relevant line in the source code, when looking at the log output in the browser tools or in VSCode. Also, for my needs, the js standard log API exposed by console is enough.

**Comment by Octodillo** [comment]
I figured I'd share my workaround if anyone else comes here disappointed. You could search and replace the different levels project wide, like `console.debug`->`//console.debug`, to enable and disable different levels. It would still be much better to have an enum value you can set on the console object though.

**Reply by mxcl** [reply]
this is a fine suggestion! an alternative people may like is: `console.debug = () => {}`

**Comment by jupegarnica** [comment]
We really need a console.setLevel() method in the standard.  And stop using third party libraries for such a basic use case.  
@lucacasonato do you know if there is something like this in the oven already? 

I am dreaming for years with it

**Comment by luolong** [comment]
Browser UI allows filtering of output based on log levels. 

I'd say, adding command line argument to have similar filtering in the CLI application makes a wolrd of sense.

**Comment by bartlomieju** [comment]
I think we _could_ add a flag that allows filtering. Before we do that, an issue with proposed flag name should be opened and agreed upon.

**Comment by jcayzac** [comment]
Right now, `.debug()`, `.info()` and `.log()` send to `1` (stdout) while `.warn()` and `.error()` send to `2` (stderr), and while they make sense they are arbitrary.

Letting us specify an append-only destination for each level on the command line, being a file handle number or an actual file path, wouldn't do any suppression by itself but offer more flexibility. People could also choose to suppress by sending those logs to /dev/null. Invoked from Bash, this would even let you stream logs over the network directly (using Bash's [builtin network support](https://who23.github.io/2020/12/03/sockets-in-your-shell.html)).
